RealmSwift/Object.swift

```diff
@@ -410,69 +410,109 @@ public class ObjectUtil: NSObject {
-            if type is Optional<String>.Type || type is Optional<NSString>.Type {
-                properties[name] = NSNumber(value: PropertyType.string.rawValue)
-            } else if type is Optional<Date>.Type {
-                properties[name] = NSNumber(value: PropertyType.date.rawValue)
-            } else if type is Optional<Data>.Type {
-                properties[name] = NSNumber(value: PropertyType.data.rawValue)
-            } else if type is Optional<Object>.Type {
-                properties[name] = NSNumber(value: PropertyType.object.rawValue)
-            } else if type is RealmOptional<Int>.Type ||
-                      type is RealmOptional<Int8>.Type ||
-                      type is RealmOptional<Int16>.Type ||
-                      type is RealmOptional<Int32>.Type ||
-                      type is RealmOptional<Int64>.Type {
-                properties[name] = NSNumber(value: PropertyType.int.rawValue)
-            } else if type is RealmOptional<Float>.Type {
-                properties[name] = NSNumber(value: PropertyType.float.rawValue)
-            } else if type is RealmOptional<Double>.Type {
-                properties[name] = NSNumber(value: PropertyType.double.rawValue)
-            } else if type is RealmOptional<Bool>.Type {
-                properties[name] = NSNumber(value: PropertyType.bool.rawValue)
-            } else if prop.value as? RLMOptionalBase != nil {
-                throwRealmException("'\(type)' is not a valid RealmOptional type.")
-            } else if mirror.displayStyle == .optional || type is ExpressibleByNilLiteral.Type {
-                properties[name] = NSNull()
+    // Reflect an object, returning only children representing managed Realm properties.
+    private static func getNonIgnoredMirrorChildren(for object: Any) -> [Mirror.Child] {
+        guard let realmObject = object as? Object else {
+            return Array(Mirror(reflecting: object).children)
+        }
+        let ignoredPropNames: Set<String> = Set(type(of: realmObject).ignoredProperties())
+        // No HKT in Swift, unfortunately
+        return Mirror(reflecting: realmObject).children.filter { (prop: Mirror.Child) -> Bool in
+            guard let label = prop.label else {
+                return false
+            }
+            if ignoredPropNames.contains(label) {
+                // Ignored property.
+                return false
             }
-            return properties
+            if let lazyBaseName = baseName(forLazySwiftProperty: label) {
+                if ignoredPropNames.contains(lazyBaseName) {
+                    // Ignored lazy property.
+                    return false
+                }
+                // Managed lazy property; not currently supported.
+                // FIXME: revisit this once Swift gets property behaviors/property macros.
+                throwRealmException("Lazy managed property '\(lazyBaseName)' is not allowed on a Realm Swift object"
```

Comment:
could you write a test confirming that this exception is thrown for non-ignored lazy properties? Could use our `FakeObject` utilities like we do for other invalid schema tests.